src/auth/webhook-secret.guard.ts

```diff
@@ -0,0 +1,32 @@
+import {
+  CanActivate,
+  ExecutionContext,
+  Injectable,
+  UnauthorizedException,
+} from '@nestjs/common';
+import * as crypto from 'crypto';
+import { Request } from 'express';
+import { Env } from 'src/env';
+
+@Injectable()
+export class WebhookSecretGuard implements CanActivate {
+  canActivate(context: ExecutionContext): boolean {
+    const request: Request = context.switchToHttp().getRequest();
+    const signature = request.headers['x-hub-signature-256'] as string;
+    const secret = Env.SECRET.WEBHOOK;
+    const payload = JSON.stringify(request.body);
+
+    if (!signature) {
+      throw new UnauthorizedException('Missing signature');
+    }
+
+    const hmac = crypto.createHmac('sha256', secret);
+    const digest = `sha256=${hmac.update(payload).digest('hex')}`;
+
+    if (!crypto.timingSafeEqual(Buffer.from(digest), Buffer.from(signature))) {
```

Comment:
PullProofAI

The comparison of the HMAC digest with the signature using `crypto.timingSafeEqual` is a good practice for security. However, in this context, it might lead to a potential timing attack vulnerability due to the direct comparison of the buffers.

#### Recommendation
Consider using a constant-time comparison function like `crypto.timingSafeEqual` or implement a custom constant-time comparison function to mitigate timing attacks.

```typescript
// Implement a constant-time comparison function
function safeCompare(a: Buffer, b: Buffer): boolean {
    if (a.length !== b.length) {
        return false;
    }

    let result = 0;
    for (let i = 0; i < a.length; i++) {
        result |= a[i] ^ b[i];
    }

    return result === 0;
}

// Use the safe comparison function
if (!safeCompare(Buffer.from(digest), Buffer.from(signature))) {
    throw new UnauthorizedException('Invalid signature');
}
```


<!-- This is an auto-generated comment by PullProofAI -->